### DISCUSSION TITLE
v7.0 Certificate verify failed: unable to get local issuer certificate

### DISCUSSION THREAD

**Discussion by mrsiano** [mitmproxy/mitmproxy#5094] [Q&A]
using client certificate failed due to local issuer certificate
/bin/mitmdump --listen-port 8156 --certs myexample.net=myexample.net

```
127.0.0.1:62950: GET https://myexample.net/html/clientcert/cert_det… HTTP/2.0
 << Certificate verify failed: unable to get local issuer certificate
```

**Comment by mhils** [comment]
If you want to supply a client certificate, you need to use `--set client_certs=...`. `--certs` is for server-side certificates.

https://docs.mitmproxy.org/stable/concepts-certificates/#using-a-client-side-certificate

**Reply by mrsiano** [reply]
Yes we try that too and this one retrives "tlsv1 alert unknown ca"
Client TLS handshake failed. The client does not trust the proxy's certificate for myexample.net (tlsv1 alert unknown ca)

```
./bin/mitmdump --listen-port 8156 --set client-certs=.cert http2=false --ssl-insecure=false
Only processing flows that match "http2=false & --ssl-insecure=false"
Proxy server listening at http://*:8156
127.0.0.1:63116: client connect
127.0.0.1:63116: server connect myexample.net:443 (****:443)
127.0.0.1:63116: Server TLS handshake failed. Certificate verify failed: unable to get local issuer certificate
127.0.0.1:63116: Unable to establish TLS connection with server (Certificate verify failed: unable to get local issuer certificate). Trying to establish TLS with client anyway.
127.0.0.1:63116: server disconnect myexample.net:443 (****:443)
127.0.0.1:63116: Client TLS handshake failed. The client does not trust the proxy's certificate for myexample.net (tlsv1 alert unknown ca)
```

**Reply by mhils** [reply]
It looks like your client does not trust mitmproxy's cert, and for the upstream connection the certificate chain is incomplete. There's not really anything mitmproxy can do about that, you need to supply the correct certs or convince client and server via some other means.

**Reply by mbatchelder-te** [reply]
```
127.0.0.1:62950: GET https://myexample.net/html/clientcert/cert_det… HTTP/2.0
 << Certificate verify failed: unable to get local issuer certificate
```
If this is the message from the server logs, then the server can't construct a chain from the client certificate to a trusted anchor (i.e. a root certificate) contained in the server's certificate store used for validating client cert chains. So check your web server's configuration/documentation.

**Comment by mrsiano** [comment]
Thanks will double check the certs

**Reply by mbatchelder-te** [reply]
Please post the full curl output with the command line.

**Reply by mrsiano** [reply]
```
curl --proxy 127.0.0.1:8156 --cacert cert.cer https://myserver.net.net/html/clientcert/cert_details_require.asp --insecure -vvv
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 8156 (#0)
* allocate connect buffer!
* Establish HTTP proxy tunnel to myserver.net.net:443
> CONNECT myserver.net.net:443 HTTP/1.1
> Host: myserver.net.net:443
> User-Agent: curl/7.64.1
> Proxy-Connection: Keep-Alive
> 
< HTTP/1.1 200 Connection established
< 
* Proxy replied 200 to CONNECT request
* CONNECT phase completed!
* ALPN, offering h2
* ALPN, offering http/1.1
* error setting certificate verify locations, continuing anyway:
*   CAfile: cert.cer
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* CONNECT phase completed!
* CONNECT phase completed!
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* ALPN, server did not agree to a protocol
* Server certificate:
*  subject: CN=myserver.net.net; O=LoadRunner
*  start date: Jan 30 10:19:21 2022 GMT
*  expire date: Feb  1 10:19:21 2023 GMT
*  issuer: CN=mitmproxy; O=mitmproxy
*  SSL certificate verify result: unable to get local issuer certificate (20), continuing anyway.
> GET /html/clientcert/cert_details_require.asp HTTP/1.1
> Host: myserver.net.net
> User-Agent: curl/7.64.1
> Accept: */*
```
and it just stuck at this point.

I also attaching the print of the mitmproxy layers for correlation
server_connect: {data}
server_connected: {data}
tls_start_server: {tls_start}
request: {flow} -> here we are get stuck

**Comment by mrsiano** [comment]
I'll try to clarify my question, is mitmproxy should support "TLSv1.2 (IN), TLS handshake, Hello request (0)" ? cause it seems like we did not catch this event @mhils @mbatchelder-te

**Comment by mrsiano** [comment]
we also see that * TLSv1.2 (IN), TLS handshake, Request CERT (13): is missing mitmproxy did not catch it too.

this is a curl log from working flow w\o mitmproxy
```
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Request CERT (13):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Certificate (11):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS handshake, CERT verify (15):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
```

**Comment by Shnitzelil** [comment]
Same here.. we have client cert issue with same server

**Comment by mrsiano** [comment]
This problem recurs when working with an IIS server but with an Apache server it works smoothly, the difference is in the IIS server there is a reconnect that triggers a reconnection at the TLS level where the certificate was sent (that event did not catch in mitmproxy level), in Apache server it is sent immediately without reconnecting

**Comment by Shnitzelil** [comment]
Hi,
Any updates?

**Comment by mhils** [comment]
I would need a simple and self-contained reproducible example to get started at all.

**Reply by mrsiano** [reply]
https://github.com/mitmproxy/mitmproxy/issues/5511
I've mentioned here how to reproduce
